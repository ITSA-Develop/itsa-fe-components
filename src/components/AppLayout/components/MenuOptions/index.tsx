import { TExtendedMenuItem, TMenuMode } from '@/types';
import { Menu, MenuProps } from 'antd';
import { SyncOutlined } from '@ant-design/icons';
import React from 'react';

export interface MenuOptionsProps {
	collapsed: boolean;
	currentPath: string;
	onClickOptionMenu: (info: { key: string; item: TExtendedMenuItem }) => void;
	items?: TExtendedMenuItem[];
	mode?: TMenuMode;
	openKeys?: string[];
	loading?: boolean;
	onOpenKeysChange?: (openKeys: string[]) => void;
}

const MenuOptions: React.FC<MenuOptionsProps> = ({
	collapsed,
	currentPath = '',
	onClickOptionMenu,
	items = [],
	mode = 'inline',
	openKeys = [],
	onOpenKeysChange,
	loading = false,
}) => {
	const handleMenuClick: MenuProps['onClick'] = info => {
		const findMenuItem = (menuItems: TExtendedMenuItem[], key: string): TExtendedMenuItem | null => {
			for (const item of menuItems) {
				if (item.key === key) {
					return item;
				}
				if ('children' in item && item.children) {
					const found = findMenuItem(item.children as TExtendedMenuItem[], key);
					if (found) return found;
				}
			}
			return null;
		};

		const clickedItem = findMenuItem(items, info.key);
		if (clickedItem && onClickOptionMenu) {
			onClickOptionMenu({ key: info.key, item: clickedItem });
		}
	};

	return (
		<div className="menu-options">
			{loading ? (
				<div className="flex flex-col items-center justify-center h-full pt-10 gap-2">
					<SyncOutlined spin className='text-gray-400' />
					<small className='text-gray-400'>Cargando men√∫</small>
				</div>
			) : (
				<Menu
					mode={mode}
					inlineCollapsed={collapsed}
					items={items}
					className="overflow-auto scrollbar-none border-none min-w-full"
					defaultSelectedKeys={[currentPath]}
					openKeys={openKeys}
					onOpenChange={onOpenKeysChange}
					onClick={handleMenuClick}
				/>
			)}
		</div>
	);
};

export default MenuOptions;
