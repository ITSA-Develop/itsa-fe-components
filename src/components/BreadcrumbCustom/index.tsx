import { Button } from 'antd';
import { LeftOutlined } from '@ant-design/icons';

export interface IBreadcrumbCustomProps {
	title: string;
	description: string;
	action: () => void;
}
export const BreadcrumbCustom = ({ title, description, action }: IBreadcrumbCustomProps) => {
	return (
		<div className="flex flex-row items-center gap-2">
			<Button onClick={action} type="text" icon={<LeftOutlined className="font-bold" style={{ fontSize: '20px' }} />} />
			<div className="flex flex-row items-center gap-2">
				<h4 className="font-bold text-lg text-black-100">{title}</h4>
				<span className="text-black-100">|</span>
				<small className="text-sm text-black-100">{description}</small>
			</div>
		</div>
	);
};
