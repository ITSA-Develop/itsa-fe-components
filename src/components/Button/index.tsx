import { ButtonAntd } from '../ButtonAntd';
import { ReactNode } from 'react';

export interface IButtonProps {
	size?: 'small' | 'middle' | 'large';
	type?: 'primary' | 'secondary';
	title?: ReactNode;
	disabled?: boolean;
	onClick?: () => void;
	/** Cuando es true en tipo secondary, aplica estilo gris (#595959) */
	default?: boolean;
}

export const Button = (props: IButtonProps) => {
	const { size = 'small', type = 'primary', title, disabled = false, onClick } = props;
	const sizeClass = size === 'small' ? 'itsa-btn--sm' : size === 'middle' ? 'itsa-btn--md' : 'itsa-btn--lg';
	const variantClass = type === 'primary' ? 'itsa-btn--primary' : 'itsa-btn--secondary';
	const disabledClass = disabled ? 'itsa-btn--disabled' : '';
	const defaultSecondaryClass = type === 'secondary' && props.default ? 'itsa-btn--default' : '';
	const className = ['itsa-btn', sizeClass, variantClass, defaultSecondaryClass, disabledClass].filter(Boolean).join(' ');
	const antdType: 'primary' | 'default' = type === 'primary' ? 'primary' : 'default';

	return (
		<ButtonAntd className={className} size={size} type={antdType} disabled={disabled} onClick={onClick}>
			{title}
		</ButtonAntd>
	);
};
