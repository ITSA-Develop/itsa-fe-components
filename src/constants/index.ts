import axios from 'axios';
import dayjs from 'dayjs';

export const DATE_FORMAT_UI = 'DD/MM/YYYY';
export const DATE_YEAR_FORMAT_UI = 'YYYY';
export const DATE_FORMAT_DAYJS_DEFAULT = 'YYYY-MM-DD';

export const TABLE_HEADER_ACTIONS = 'actions';

const AXIOS_INITIAL_CONF = {
	headers: {
		'Content-Type': 'application/json',
		Accept: 'application/json',
	},
};

export const instance = axios.create({
	...AXIOS_INITIAL_CONF,
});

export const REGEX_FORMATS = {
	email: /^(?!.*\.\.)[a-zA-Z0-9](\.?[a-zA-Z0-9_%+-]+)*@[a-zA-Z0-9-]+(\.[a-zA-Z0-9-]+)*\.[a-zA-Z]{2,}$/,
};

export const MAX_DATE_ADULT = dayjs().subtract(18, 'year');
export const TODAY = dayjs();

export const FILE_TYPES = {
	excel: '.xls,.xlsx,application/vnd.ms-excel,application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
	word: '.doc,.docx,application/msword,application/vnd.openxmlformats-officedocument.wordprocessingml.document',
	text: 'text/plain',
	pdf: 'application/pdf',
	zip: '.zip,application/zip',
};

export const ROWS_PER_PAGE_DEFAULT = 10;
export const ROWS_PER_PAGE = [10, 25, 50, 100];

export const ROUTES_IMAGES = {
	companyLogo: 'src/assets/images/image-itsa-logo.png',
};

export const LOGO_DIMENSIONS = {
	HEADER_WIDTH: 150.46,
	HEADER_HEIGHT: 24,
};

export const DEFAULT_PAGINATION_CONFIG = {
	current: 1,
	pageSize: 20,
	showSizeChanger: true,
	showQuickJumper: false,
	showTotal: (total: number) => `Total ${total} items`,
	pageSizeOptions: [10, 20, 50, 100],
	hideOnSinglePage: false,
	showLessItems: true,
};

export const TABLE_SCROLL = {
	y: 300,
};

export * from './agencies';

export const API_VERSIONS_MODULE = {
	security: 'sec/v1/security/',
};

export const TABS_ITEM_CONTENT_WIDTH = '512px';

export const COLOR_TAGS = {
	magenta: {
		text: '#EB2F96',
		borde: '#FFADD2',
		bg: '#FFF0F6',
	},
	green: {
		text: '#52A927',
		borde: '#9FCA7F',
		bg: '#F6FFED',
	},
	blue: {
		text: '#2F54EB',
		borde: '#ADC6FF',
		bg: '#F0F5FF',
	},
	orange: {
		text: '#FA8C16',
		borde: '#FFD591',
		bg: '#FFF7E6',
	},
};