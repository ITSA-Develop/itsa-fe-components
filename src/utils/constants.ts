import { IId } from '@/interfaces';
import * as mdiIcons from '@mdi/js';
import { PermissionAction } from '../types';

// TODO: double check with BE what is going to be the Iid returned by default.
export const excludedIIdFields: (keyof IId)[] = ['id', 'modified_by', 'created_by', 'modified_on', 'created_on'];

export const sessionExpiredQueryParam = 'sessionExpired';
export const newPasswordRequiredQueryParam = 'newPasswordRequired';

export const BASE_COUNTRY = 'ECUADOR';

export const productionDomainLink = 'https://itsa-tomebamba.com'; // TODO: TBD

export const GOOGLE_MAP_ADDRESS_KEYS = {
	streetNumber: 'street_number',
	route: 'route',
	postalCode: 'postal_code',
	province: 'administrative_area_level_1',
	canton: 'administrative_area_level_2',
	parish: 'sublocality_level_1',
	lat: 'lat',
	long: 'long',
};

export const GOOGLE_API_KEY = import.meta.env.VITE_GOOGLE_MAPS_API_KEY;
export const ENVIRONMENT = import.meta.env.VITE_ENV;

export const REQUIRED = { required: 'Requerido' };

export const UNKNOWN_ERROR = 'Error desconocido';

export const ICON_MAP = mdiIcons;

export const PERMISSIONS_ACTIONS: Record<PermissionAction, PermissionAction> = {
	create: 'create',
	read: 'read',
	update: 'update',
	delete: 'delete',
	all_actions: 'all_actions',
};

export const SUBDOMAINS = {
	backOffice: 'backoffice',
	frontOffice: 'frontoffice',
	core: 'core',
};

export const AUTH_LINK = 'http://192.168.7.87:8082/'; // TODO: update this when they have clear the domain

export const URL_PARAMS = {
	redirect: 'redirect',
};
