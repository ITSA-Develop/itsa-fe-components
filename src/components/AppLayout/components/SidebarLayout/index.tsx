import MenuOptions from '@/components/AppLayout/components/MenuOptions';
import { getAllMenuKeys, transformModuleToMenuData } from '@/helpers';
import { filterMenuItems } from '@/helpers/menu/menuDataTransformer';
import { useSidebarStore } from '@/hooks';
import { useAppLayoutStore } from '@/store/appLayout.store';
import { TExtendedMenuItem, TMenuMode } from '@/types';
import { DoubleLeftOutlined, SearchOutlined } from '@ant-design/icons';
import { Button, Input, Layout } from 'antd';
import { Content } from 'antd/es/layout/layout';
import { ReactNode, useEffect, useMemo } from 'react';
export interface SidebarLayoutProps {
	children: ReactNode;
	onClickOptionMenu: (info: { key: string; item: TExtendedMenuItem }) => void;
	width?: number;
	currentPath?: string;
	modeSidebar?: TMenuMode;
}
export const SidebarLayout = ({
	children,
	onClickOptionMenu,
	width = 235,
	currentPath = '',
	modeSidebar = 'inline',
}: SidebarLayoutProps) => {
	const currentModule  = useAppLayoutStore((state) => state.currentModule);

	const { collapsed, setCollapsed, searchTerm, setSearchTerm, openKeys, setOpenKeys } = useSidebarStore();
	const menuData = useMemo(() => {
		if (!currentModule) {
			return [];
		}
		const originalMenuData = transformModuleToMenuData(currentModule);

		if (!searchTerm.trim()) {
			return originalMenuData;
		}

		return filterMenuItems(originalMenuData, searchTerm);
	}, [currentModule, searchTerm]);

	useEffect(() => {
		if (searchTerm.length > 0) {
			const allKeys = getAllMenuKeys(menuData);
			setOpenKeys(allKeys);
		} else {
			setOpenKeys([]);
		}
	}, [searchTerm, menuData, setOpenKeys]);

	return (
		<Layout hasSider className="gap-2">
			{!collapsed && (
				<div className="flex flex-col pt-3 rounded-lg bg-gray-200" style={{ width: width }}>
					<div className="flex items-center justify-center pr-3 pl-3">
						<Input
							placeholder="Buscar en el menÃº"
							className="rounded-lg text-sm"
							suffix={<SearchOutlined className="text-gray-300" />}
							defaultValue={searchTerm}
							onChange={e => setSearchTerm(e.target.value)}
						/>
					</div>
					<div className="flex-1 overflow-y-auto scrollbar-none h-full max-w-full">
						<MenuOptions
							items={menuData}
							collapsed={collapsed}
							onClickOptionMenu={onClickOptionMenu}
							currentPath={currentPath || ''}
							mode={modeSidebar}
							openKeys={openKeys}
							onOpenKeysChange={setOpenKeys}
						/>
					</div>
					<div className="w-full flex justify-end pr-3 pl-3">
						<Button
							type="link"
							onClick={() => setCollapsed(true)}
							icon={<DoubleLeftOutlined className="text-gray-400" />}
						/>
					</div>
				</div>
			)}
			<Layout className="rounded-lg">
				<Content className="bg-white-100 overflow-auto rounded-lg pt-3 pb-2 pr-4 pl-4">{children}</Content>
			</Layout>
		</Layout>
	);
};
