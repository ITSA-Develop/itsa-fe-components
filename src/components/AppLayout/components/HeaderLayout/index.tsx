import { ActiveNotificationIcon, NotificationIcon, PinIcon, UserIcon } from '@/assets/icons';
import { DropdownCustomLabel } from '@/components/DropdownCustomLabel';
import { DropdownIcon } from '@/components/DropdownIcon';
import { Image } from '@/components/Image';
import { LOGO_DIMENSIONS, ROUTES_IMAGES } from '@/constants';
import { useSidebarStore } from '@/hooks';
import { ISelectOptionDropdownButton } from '@/interfaces';
import { useAppLayoutStore } from '@/store/appLayout.store';
import { MenuUnfoldOutlined, SettingOutlined } from '@ant-design/icons';
import { Button, Dropdown, MenuProps } from 'antd';

export interface HeaderLayoutProps {
	loadingHeader: boolean;
	notifications?: MenuProps;
	userActions?: MenuProps;
	logo?: string;
}

export const HeaderLayout = ({
	loadingHeader,
	notifications = { items: [] },
	userActions = { items: [] },
	logo = '',
}: HeaderLayoutProps) => {
	const { setCurrentModule, setCurrentAgency } = useAppLayoutStore();
	const { collapsed, setCollapsed } = useSidebarStore();
	const { agencies } = useAppLayoutStore();
	const { currentAgency } = useAppLayoutStore();
	const { currentModule } = useAppLayoutStore();
	const { modulesAgency } = useAppLayoutStore();
	const { userName } = useAppLayoutStore();
	const { userRole } = useAppLayoutStore();

	const isActiveNotifications = Boolean(notifications?.items?.length);
	const isActiveUserActions = Boolean(userActions?.items?.length);

	const modulesData: ISelectOptionDropdownButton = {
		items: modulesAgency.map(m => ({
			key: m.id.toString(),
			label: m.name,
			onClick: (id: string) => handleSetCurrentModule(id),
		})),
	};

	const agenciesData: ISelectOptionDropdownButton = {
		items: agencies.map(a => ({
			key: a.id.toString(),
			label: a.name,
			onClick: (id: string) => handleSetCurrentAgency(id),
		})),
	};

	const handleSetCurrentModule = (moduleId: string) => {
		const module = modulesAgency.find(m => m.id.toString() === moduleId);
		if (module) {
			setCurrentModule(module);
		}
	};

	const handleSetCurrentAgency = (agencyId: string) => {
		const agency = agencies.find(a => a.id.toString() === agencyId);
		if (agency) {
			setCurrentAgency(agency);
		}
	};

	return (
		<header className="h-16">
			<div className="flex flex-row text-white-100 items-center justify-between w-full rounded-xl h-16 pr-4 pl-6 bg-primary-700">
				<div className="w-full flex flex-row items-center justify-start gap-4">
					{collapsed && (
						<Button
							type="text"
							icon={<MenuUnfoldOutlined className="text-white-100" />}
							onClick={() => setCollapsed(!collapsed)}
						/>
					)}
					<Image
						imgPath={logo || ROUTES_IMAGES.companyLogo}
						width={LOGO_DIMENSIONS.HEADER_WIDTH}
						height={LOGO_DIMENSIONS.HEADER_HEIGHT}
						alt="Logo"
						errorLabel="Importadora Tomebamba"
					/>
				</div>
				<div className="hidden sm:flex flex-row w-auto items-center gap-4">
					<div className="flex md:hidden flex-row items-center gap-4">
						<DropdownIcon
							options={modulesData}
							loading={loadingHeader}
							icon={<SettingOutlined className="text-white-100 w-4 h-4" />}
							onChange={handleSetCurrentModule}
						/>
						<DropdownIcon
							options={agenciesData}
							loading={loadingHeader}
							icon={<PinIcon className="text-white-100 w-4 h-4" />}
							onChange={handleSetCurrentAgency}
						/>
					</div>

					<div className="hidden md:flex flex-row items-center gap-4">
						<DropdownCustomLabel
							defaultValue={currentModule?.id?.toString()}
							options={modulesData}
							loading={loadingHeader}
							icon={<SettingOutlined className="text-white-100 w-4 h-4" />}
							emptyLabel="Sin mÃ³dulos asignados"
							onChange={handleSetCurrentModule}
						/>
						<DropdownCustomLabel
							defaultValue={currentAgency?.id?.toString()}
							options={agenciesData}
							loading={loadingHeader}
							icon={<PinIcon className="text-white-100 w-4 h-4" />}
							emptyLabel="Sin agencias asignadas"
							onChange={handleSetCurrentAgency}
						/>
						<div className="flex flex-col">
							<span className="text-4 whitespace-nowrap">{userName ?? ''}</span>
							<span className="text-primary-900 font-bold text-end text-xs whitespace-nowrap">{userRole ?? ''}</span>
						</div>
					</div>
				</div>
				<div className="flex flex-row pl-4">
					<Dropdown menu={userActions || { items: [] }} placement="bottomRight" disabled={!isActiveUserActions}>
						<Button type="text" icon={<UserIcon className="text-white-100 w-6 h-6" />} />
					</Dropdown>
					<Dropdown menu={notifications || { items: [] }} placement="bottomRight" disabled={!isActiveNotifications}>
						<Button
							type="text"
							icon={
								isActiveNotifications ? (
									<ActiveNotificationIcon className="text-white-100 w-6 h-6" />
								) : (
									<NotificationIcon className="text-white-100 w-6 h-6" />
								)
							}
						/>
					</Dropdown>
				</div>
			</div>
		</header>
	);
};
