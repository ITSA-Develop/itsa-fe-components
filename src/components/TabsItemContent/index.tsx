import { TABS_ITEM_CONTENT_WIDTH } from '@/constants';
import { Typography } from 'antd';
import { ReactNode } from 'react';

const { Text } = Typography;

export interface ITabsItemContent {
	children: ReactNode;
	width?: string;
	title?: string;
}

export const TabsItemContent = ({ children, width = TABS_ITEM_CONTENT_WIDTH, title }: ITabsItemContent) => {
	return (
		<div className="flex flex-col justify-center items-center h-full pb-4">
			{title && (
				<div className="flex justify-start items-center w-full pb-2" style={{ width: width }}>
					<Text className="text-sm font-bold">{title}</Text>
				</div>
			)}
			<div className="flex flex-col justify-center items-center h-full rounded-lg bg-gray-250" style={{ width: width }}>
				{children}
			</div>
		</div>
	);
};
