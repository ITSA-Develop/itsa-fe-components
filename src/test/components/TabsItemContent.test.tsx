import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { TabsItemContent } from '../../components/TabsItemContent';
import { TABS_ITEM_CONTENT_WIDTH } from '../../constants';

describe('TabsItemContent component', () => {
    it('renders with title and children', () => {
        const { container } = render(
            <TabsItemContent title="Mi Título">
                <div>Contenido</div>
            </TabsItemContent>,
        );

        expect(screen.getByText('Mi Título')).toBeInTheDocument();
        expect(screen.getByText('Contenido')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('applies default and custom width', () => {
        const { rerender } = render(
            <TabsItemContent>
                <div>Default</div>
            </TabsItemContent>,
        );

        const defaultContainer = screen.getByText('Default').parentElement as HTMLElement;
        expect(defaultContainer).toHaveStyle({ width: TABS_ITEM_CONTENT_WIDTH });

        rerender(
            <TabsItemContent width="320px">
                <div>Custom</div>
            </TabsItemContent>,
        );
        const customContainer = screen.getByText('Custom').parentElement as HTMLElement;
        expect(customContainer).toHaveStyle({ width: '320px' });
    });
});


