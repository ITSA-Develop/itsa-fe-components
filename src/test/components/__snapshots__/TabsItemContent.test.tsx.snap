// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TabsItemContent component > renders with title and children 1`] = `
<div>
  <div
    class="flex flex-col justify-center items-center h-full pb-4"
  >
    <div
      class="flex justify-start items-center w-full pb-2"
      style="width: 512px;"
    >
      <span
        class="ant-typography text-sm font-bold css-dev-only-do-not-override-l9pxc0"
      >
        Mi TÃ­tulo
      </span>
    </div>
    <div
      class="flex flex-col justify-center items-center h-full rounded-lg bg-gray-250"
      style="width: 512px;"
    >
      <div>
        Contenido
      </div>
    </div>
  </div>
</div>
`;
